### PR TITLE
Add support for multiple hashing algorithms

### LINKED ISSUE #108: Add support for multiple hash algorithms in artifacts
Currently we support only md5, but we should also support sha1 and sha256.

### PR DESCRIPTION
**Pull request by goldmann**
Fixes #108.

### CODE REVIEW

File: dogen/generator.py [hunk 1]
```diff
@@ -289,11 +291,27 @@ def handle_sources(self):
...
             filename = ("%s/%s" % (self.output, target))
+
             passed = False
+            algorithm = None
+
+            md5sum = source.get('md5sum')
+
+            # Backwards compatibility for md5sum
+            if md5sum and not source.get('md5'):
+                self.log.warn("The 'md5sum' key is deprecated, please use 'md5' for %s. Or better switch to sha1 or sha256." % url)
```
Comment (l-d-j) [inline dogen/generator.py]: we should warn about deprecation of md5sum regardless of md5 key being present

File: dogen/generator.py [hunk 2]
```diff
@@ -289,11 +291,27 @@ def handle_sources(self):
...
+                source['md5'] = md5sum
+
+            for supported_algorithm in SUPPORTED_HASH_ALGORITHMS:
+                if not source.get(supported_algorithm):
+                    continue
+
+                algorithm = supported_algorithm
```
Comment (l-d-j) [inline dogen/generator.py]: so if i specify more alg is it first match? in that case we should check all.

File: dogen/generator.py (hunk 1 above)
Comment (goldmann) [inline dogen/generator.py]: Good catch!

File: dogen/generator.py (hunk 2 above)
Comment (goldmann) [inline dogen/generator.py]: Yes. Do you think we need to change it? We select the most secure one. Isn't sufficient? Or maybe a warning? WDYT?

Comment (l-d-j) [inline dogen/generator.py]: if we let users to specify more - we should check all of them - what the point of choosing - if user wants to check only one - he will specify only one, if he specify more of them - check them all.

**Comment by goldmann** [comment]
I've updated this PR. The rules are now following:

1. If multiple algorithms are specified - all hashes are checked when the files is downloaded.
2. If the `DOGEN_SOURCES_CACHE` env variable is set, only the most secure hash algorithm is used to generate the url: sh256 > sha1 > md5.
3. Only most secure hash algorithm is used to construct the `artifacts` dict later used by the plugins.

**Pull request merged**